sklearn/ensemble/_hist_gradient_boosting/histogram.pyx

```diff
@@ -33,8 +36,9 @@ from .common cimport G_H_DTYPE_C
 #       array[i + 2] = i + 2
 #       array[i + 3] = i + 3
 #
-#   This is to hint gcc that it can auto-vectorize these 4 operations and
-#   perform them all at once.
+#   This is to hint gcc that it can use instruction-level parallelism for these 4
+#   operations and perform them all at once. As 2 points may both update the same
+#   bin, SIMD is not possible here.
```

Comment:
That's right. Moreover, we use Arrays of Structures (namely arrays of `hist_struct`) for histogram, so SIMD is not possible even if they were to update different values.
```suggestion
#   operations and perform them all at once. As 2 points may both update the same
#   bin and updated values aren't contiguous in memory, SIMD is impossible here.
```

Naive question: could using Structures of Arrays for histograms help with increasing parallelization (both at the level of instructions and with SIMD) and reducing false sharing?

Is there some design decisions in this regards, @NicolasHug? :slightly_smiling_face:

Reply:
If SIMD isn't possible, what is that "instruction-level parallelism" that's happening then? (I thought that was SIMD.)

> Is there some design decisions in this regards, @NicolasHug? 🙂

IIRC the "design decision" was that we had one implementation for histogram building, saw that light-GBM did it this way (the way it's currently done on `main`), and we observed massive speed-up when we replicated that. As for array of struct vs struct of arrays: I didn't consider the alternative (I was struggling enough juggling with a bunch of undocumented Cython features).

Basically, anything you can think of is probably worth exploring